File: src/main/scala/com/xmlcalabash/steps/os/OsInfo.scala
package com.xmlcalabash.steps.os

import com.xmlcalabash.model.util.{SaxonTreeBuilder, XProcConstants}
import com.xmlcalabash.runtime.{StaticContext, XProcMetadata, XmlPortSpecification}
import com.xmlcalabash.steps.DefaultXmlStep
import com.xmlcalabash.util.{MediaType, TypeUtils}
import net.sf.saxon.om.{AttributeMap, EmptyAttributeMap}
import net.sf.saxon.s9api.QName

import scala.collection.mutable
import scala.jdk.CollectionConverters.MapHasAsScala

class OsInfo extends DefaultXmlStep {
  // I should make this static...
  private val nameMap = mutable.HashMap.empty[String, String]
  nameMap.put("file.separator", "file-separator")
  nameMap.put("path.separator", "path-separator")
  nameMap.put("os.arch", "os-architecture")
  nameMap.put("os.name", "os-name")
  nameMap.put("os.version", "os-version")
  nameMap.put("user.dir", "cwd")
  nameMap.put("user.name", "user-name")
  nameMap.put("user.home", "user-home")

  override def inputSpec: XmlPortSpecification = XmlPortSpecification.NONE

  override def outputSpec: XmlPortSpecification = XmlPortSpecification.XMLRESULT

  override def run(context: StaticContext): Unit = {
    super.run(context)

    var amap: AttributeMap = EmptyAttributeMap.getInstance()
    for (property <- System.getProperties.asScala.keySet) {
      val pname = property.toString
      val value = System.getProperty(pname)
      try {
        val name = if (nameMap.contains(pname)) {
          new QName("", nameMap(pname))
        } else {
          // Substituting "-" for "." seems to make the list more consistent
          new QName("", pname.replace(".", "-"))
        }
        amap = amap.put(TypeUtils.attributeInfo(name, value))
      } catch {
        case _: Exception =>
          // Nevermind, if the property name can't be a QName, I give up
          ()
      }
    }

    val builder = new SaxonTreeBuilder(config)
    builder.startDocument(None)
    builder.addStartElement(XProcConstants.c_result, amap)

    for (name <- System.getenv().asScala.keySet) {
      val value = System.getenv(name)
      var emap: AttributeMap = EmptyAttributeMap.getInstance()
      emap = emap.put(TypeUtils.attributeInfo(XProcConstants._name, name))
      emap = emap.put(TypeUtils.attributeInfo(XProcConstants._value, value))
      builder.addStartElement(XProcConstants.c_environment, emap)
      builder.addEndElement()
    }

    builder.addEndElement()
    builder.endDocument()
    consumer.get.receive("result", builder.result, new XProcMetadata(MediaType.XML))
  }
}

File: xmlcalabash/src/main/kotlin/com/xmlcalabash/steps/os/OsInfo.kt
package com.xmlcalabash.steps.os

import com.xmlcalabash.documents.XProcDocument
import com.xmlcalabash.namespace.Ns
import com.xmlcalabash.namespace.NsC
import com.xmlcalabash.namespace.NsCx
import com.xmlcalabash.steps.AbstractAtomicStep
import com.xmlcalabash.util.SaxonTreeBuilder
import net.sf.saxon.om.AttributeMap
import net.sf.saxon.om.EmptyAttributeMap
import net.sf.saxon.s9api.QName
import org.apache.logging.log4j.kotlin.logger

class OsInfo(): AbstractAtomicStep() {
    companion object {
        private val mapping = mapOf(
            "file.separator" to "file-separator",
            "path.separator" to "path-separator",
            "os.arch" to "os-architecture",
            "os.name" to "os-name",
            "os.version" to "os-version",
            "user.dir" to "cwd",
            "user.name" to "user-name",
            "user.home" to "user-home"
        )
    }

    var onlyStandardProperties = false

    override fun input(port: String, doc: XProcDocument) {
        // none
    }

    override fun extensionAttributes(attributes: Map<QName, String>) {
        super.extensionAttributes(attributes)
        val value = attributes[NsCx.onlyStandard]
        if (value != null) {
            if (value == "true" || value == "false") {
                onlyStandardProperties = value == "true"
            } else {
                logger.debug { "Ignoring unexpected value for cx:only-standard: ${value}"}
            }
        }
    }

    override fun run() {
        super.run()

        val attr = mutableMapOf<QName, String?>()
        for ((pname, pvalue) in System.getProperties()) {
            val name = pname.toString()
            val value = pvalue.toString()
            if (mapping.contains(name)) {
                attr[QName(mapping[name])] = value
            } else if (!onlyStandardProperties) {
                val qname = QName(NsCx.namespace, "cx:${name.replace(".", "-")}")
                attr[qname] = value
            }
        }

        val builder = SaxonTreeBuilder(stepConfig)
        builder.startDocument(null)
        builder.addStartElement(NsC.result, stepConfig.attributeMap(attr))

        for ((name, value) in System.getenv()) {
            builder.addStartElement(NsC.environment, stepConfig.attributeMap(mapOf(
                Ns.name to name,
                Ns.value to value
            )))
            builder.addEndElement()
        }

        builder.addEndElement()
        builder.endDocument()

        receiver.output("result", XProcDocument.ofXml(builder.result, stepConfig))
    }

    override fun toString(): String = "p:os-info"
}